refactor user-space RDMA driver

ruiqurm opened the pull request.

EAimTY committed:
Add control ringbuf descriptor types

EAimTY committed:
Add work ringbuf descriptor types

EAimTY committed:
Update desc fields

EAimTY committed:
Expose descriptor fields

EAimTY committed:
Make device initialization fallible

EAimTY committed:
derive `Debug`, `Clone` for several enums

EAimTY committed:
implement control path verbs

EAimTY committed:
expose public structs under `qp` module

EAimTY committed:
check associated QP before destroying PD

EAimTY committed:
implement send and recv interface

EAimTY committed:
implement polling logic

EAimTY committed:
fix send report handling

EAimTY committed:
add packet receiving completion check

EAimTY committed:
use global incremental id as ctrl op id

EAimTY committed:
hide magic numbers behind constants

EAimTY committed:
correctly calculate MR key

EAimTY committed:
fix `QP_AVAILABLITY` initialization

EAimTY committed:
fix incorrect ctrl descriptor content

EAimTY committed:
implement page table

EAimTY committed:
rename send and recv function

EAimTY committed:
reply ACK message

EAimTY committed:
add device type `software`

EAimTY committed:
open shared memory in emulated device

myrfy001 committed:
add rpc client and shared memory allocator

EAimTY committed:
simplify emulator rpc client

EAimTY committed:
move device constants to parent module

EAimTY committed:
add new descriptor abstraction

EAimTY committed:
refactor ringbuf and move it to the parent module

EAimTY committed:
adapt new descriptor types

EAimTY committed:
control path descriptor serialization / deserialization

EAimTY committed:
read / write control ringbuf on device

EAimTY committed:
add example `emulated`

EAimTY committed:
implement write for `ToCardWorkRbDesc`

EAimTY committed:
implement read for `ToHostWorkRbDesc`

EAimTY committed:
implement push and pop for work ringbuf in emulated device

EAimTY committed:
aadd `get_phys_addr` in trait `DeviceAdaptor`

EAimTY committed:
fix address calc for page table descriptor

EAimTY committed:
implement ringbuf interface directly on `EmulatedDevice`

EAimTY committed:
update rpc server addr in emulated device example

EAimTY committed:
remove dead code

myrfy001 committed:
fix bug and MR/QP create call works with emulator

EAimTY committed:
implement verb-level write API

myrfy001 committed:
fix more bug and can send work queue entry now

myrfy001 committed:
add some comment

myrfy001 committed:
fix bug and simple Write request can work now

EAimTY committed:
optimize ringbuf structure

EAimTY committed:
support RDMA read operation

myrfy001 committed:
read+write now works

myrfy001 committed:
refacetor example to support two nic for debug

ruiqurm committed:
squash into one commit

ruiqurm committed:
update descriptor

ruiqurm committed:
remove .vscode

ruiqurm committed:
use `bitfield` to refactor descriptor writing and reading

ruiqurm committed:
remove some `Copy` Trait

ruiqurm committed:
refactor: add the `responser` thread

the `responser` thread is used to sent the `ack`, `nack` and `read response` automaticlly

myrfy001 commented on src/device/software/logic.rs:
```diff
@@ -104,6 +104,7 @@ impl BlueRDMALogic {
     }
 
     /// Convert a `ToCardWorkRbDesc` to a `RdmaMessage` and call the `net_send_agent` to send through the network.
+    #[allow(unreachable_patterns)]
```
1. Add a TODO here to notice you remove this attribute later.
2. This function is too long, try split it.

ruiqurm replied:
ok. I may try to split after refactoring.

myrfy001 commented on src/device/software/packet_processor.rs:
```diff
@@ -78,6 +78,7 @@ impl PacketProcessor {
         }
     }
 
+    #[allow(unreachable_patterns)]
```
Same as above

myrfy001 commented on src/poll/responser.rs:
```diff
@@ -0,0 +1,666 @@
+};
+
+/// A single MR that contains a free list of packet buffer. Each slot is 64 Byte.
+pub(crate) const ACKNOWLEDGE_BUFFER_SLOT: usize = 64;
```
use _CNT as suffix to explicitly show it's a number, and the comment in the above line is 64, but the bellow line is 1024.

if the const's name is clear, no comment is needed.

ruiqurm replied:
Got it.

ruiqurm replied:
And I think it may be better to move the constant inside the `AcknowledgeBuffer`.

myrfy001 commented on src/poll/responser.rs:
```diff
@@ -0,0 +1,666 @@
+pub(crate) const ACKNOWLEDGE_BUFFER_SIZE: usize = ACKNOWLEDGE_BUFFER_SLOT * 1024;
+
+/// A interface that allows DescResponser to push the work descriptor to the device
+pub trait WorkDescriptorSender: Send + Sync {
```
A "Sender" trait  defined in a module under "poll" is confusing. Poll should be a kind of action related to Receive, not send

ruiqurm replied:
Yes. The `responser.rs` is responsible for sending the response. I wonder if should I move this mod(`responser.rs`) outside the `poll ' directory?

myrfy001 replied:
yes, try find a new place for them.

myrfy001 commented on src/poll/responser.rs:
```diff
@@ -0,0 +1,666 @@
+    fn push(&self, desc_builder: ToCardWorkRbDescBuilder) -> Result<(), Error>;
+}
+
+impl WorkDescriptorSender for Device {
```
Did we implement too many functions for "Device"? What does a "push" action on Device mean?

And, now that Device already has a "send_work_desc", why need another trait simply do the same thing?

ruiqurm replied:
To be honest, the first reason I created the interface for the `Device` is because it makes it easier to test the single module.  Another reason is that letting the `send_work_desc` as `pub` gives other structs the ability to visit that interface.  So I wonder should we separate it as an independence trait?

myrfy001 replied:
I mean can you define `send_work_desc` in a trait so that we need only one trait?

ruiqurm replied:
Got it. I will move the method `send_work_desc` to the `WorkDescriptorSender`.

myrfy001 commented on src/poll/responser.rs:
```diff
@@ -0,0 +1,666 @@
+        device: Arc<dyn WorkDescriptorSender>,
+        recving_queue: std::sync::mpsc::Receiver<RespCommand>,
+        ack_buffers: AcknowledgeBuffer,
+        qp_table: Arc<RwLock<HashMap<u32, QpContext>>>,
```
use newtype to define u32 as a new type to represent QPN will be better. QPN is 24bits, and infact our hardware can't support that much, maybe 128 is enough, so when creating a QPN, we should check it if it exceed max limit.

ruiqurm replied:
Yes. I have replaced it with a new type. See the next commit.

myrfy001 commented on src/poll/responser.rs:
```diff
@@ -0,0 +1,666 @@
+        }
+    }
+
+    pub fn free(&self, buf: &'static mut Slot) {
```
shoud we use drop instead of free?

ruiqurm replied:
If we implement a `drop` for the `Slot`, every slot will be freed after we send it to the `descriptor builder`. However, there is still a chance that the hardware does not consume the buffer while it has been used by other packets.

The design of the queue is a FIFO structure. The used buffer will be put at the end of the queue. Since we know the speed of hardware consumption is much faster than the feeding speed of the software, and we have a large enough queue so that we can guarantee that the recently used buffer will not be used again.  It makes sense.

I will update the design above with `Drop` and let it free after the thread sends the `descriptor`

myrfy001 replied:
if `free` not take ownership, then it can be freed many times?

ruiqurm replied:
Yes, you are right.  But it seems that I can not free the slot with only `Drop` in the current situation.
Recycling the slot requires a reference to the allocator's free list.  Maybe I should add an `Arc` pointer inside the `Slot` struct so that when the slot is dropped, it automatically adds itself to the free list?

myrfy001 commented on src/poll/responser.rs:
```diff
@@ -0,0 +1,666 @@
+    }
+}
+
+type Slot = [u8];
```
should we make Slot an owned type instead of reference?

ruiqurm replied:
All the slots are references to [u8], you can refer to `AcknowledgeBuffer::alloc`. But it seems it will be confusing to name the `slot` as `[u8]` rather than `&'static mut Slot`. I will update it soon.

myrfy001 reviewed (commented):
please fix some issue and discuess some questions

ruiqurm committed:
refactor the types,mods and fix some bugs.

1. replace the u32 as `Qp`, `Psn` etc.
2. move `responder.rs` to src/
3. add `responder` in the `device`
4. fix some bugs in previous serialization.(some endian problems)
5. Separte some functions into other mods rather than the `lib` module.
6. Clean some warnings, `TODO` and `FIXME` in the code.
7. Update code according to last commit comments.

ruiqurm committed:
refactor ringbuf

ruiqurm committed:
refactor the descriptor polling thread.

ruiqurm committed:
replace `WrapRefCell` struct with `Mutex`.

ruiqurm committed:
refactor check_recv_pkt_comp_thread

ruiqurm committed:
refactor operation context

ruiqurm committed:
add write/read op context handling.
Fix a bug of ack/nack packet sending.

ruiqurm committed:
use `OnceLock` to wrap Payload"

rogercloud commented on Cargo.toml:
```diff
@@ -5,3 +5,25 @@ edition = "2021"
 [dependencies]
 thiserror = "1.0.56"
+num_enum = "0.7.2"
+socket2 = { version = "0.4", features = ["all"] }
```
Why are the `socket2` and `serial_test` not up to date? Especially `serial_test`'s latest version is `3.0.0`, current version is only `0.5`.

ruiqurm replied:
You are right. I have fixed this.

rogercloud commented on src/lib.rs:
```diff
@@ -1,51 +1,364 @@
 
-        dev
+impl WorkDescriptorSender for Device {
+    fn send_work_desc(&self, desc_builder: ToCardWorkRbDescBuilder) -> Result<(), Error> {
```
Why not just passing in the builder result?

ruiqurm replied:
At first, I thought the device might update some information about the descriptor builder. For example, the `PSN`. It seems wired currently. I have used the descriptor instead.

ruiqurm committed:
fix some bug of serialization

ruiqurm committed:
fix error layout of `SendQueueReqDescSeg1`

myrfy001 committed:
add partial function to network param config

ruiqurm committed:
fix some bugs of serialization

rogercloud commented on src/utils.rs:
```diff
@@ -0,0 +1,36 @@
+#[inline]
+fn get_first_packet_length(va: u64, pmtu: u32) -> u32 {
+    let offset = va % pmtu as u64;
+    if offset == 0 {
```
I think the if branch is not necessary. If offset == 0, `pmtu - offset` == `pmtu`

ruiqurm replied:
Yes. It's redundant. I have fixed this.

rogercloud commented on src/utils.rs:
```diff
@@ -0,0 +1,36 @@
+use crate::types::Pmtu;
+
+#[inline]
+fn get_first_packet_length(va: u64, pmtu: u32) -> u32 {
```
Actually this function return the **Max** first packet length. Please put the word "max" into the function name.

rogercloud commented on src/lib.rs:
```diff
@@ -1,51 +1,364 @@
-use crate::device::{DeviceAdaptor, EmulatedDevice, HardwareDevice};
-use std::{sync::Arc, thread};
+use crate::{
```
clippy rules should be added in the lib.rs file.

ruiqurm replied:
I have added some lints but more need to be added. After I clean the `unwrap` and `as` conversion, I may update a new version tommorow. But it may need to take another couple of days to deal with the other clippy warnings.

ruiqurm replied:
I pushed a couple of commits last week, where most of the clippy rules are added.

ruiqurm committed:
add network register

rogercloud commented on src/types.rs:
```diff
@@ -0,0 +1,275 @@
+use bitflags::bitflags;
```
this file assume the host platform is little ending, it's correct for X86 but not for all. Shall we also consider big ending platform?

ruiqurm replied:
Yes, but maybe later. I will add a `TODO:` mark here.

rogercloud commented on src/types.rs:
```diff
@@ -0,0 +1,275 @@
+        if self.0 > rhs {
+            Self(self.0 - rhs)
+        } else {
+            Self(Self::MAX - rhs + self.0)
```
What if `rhs` is larger than `Self::MAX`?

ruiqurm replied:
Yes. It could happen. I have added a modulo operation at the front of the function.

rogercloud commented on src/responser.rs:
```diff
@@ -0,0 +1,684 @@
+
+/// A thread that is responsible for sending the response to the other side
+pub(crate) struct DescResponser {
+    _thread: std::thread::JoinHandle<()>,
```
Shall we implement Drop trait to handle the thread join handler?

ruiqurm replied:
Yes. I will add a drop trait here in the next commit.

rogercloud commented on src/responser.rs:
```diff
@@ -0,0 +1,684 @@
+                            (dst_ip, common)
+                        }
+                        None => {
+                            eprintln!("Failed to get QP from QP table: {:?}", ack.dpqn);
```
Can we replace this kind of `eprint` with `log::error()`?

ruiqurm replied:
Yes. I have added the `clippy::print_stderr`, `clippy::print_stdout` lints and replaced them as log::error! etc.

rogercloud reviewed (commented):
too many unwraps. please add lint and clippy ASAP.

ruiqurm committed:
fix some logic bugs.

ruiqurm committed:
test read

ruiqurm committed:
temp commit

ruiqurm committed:
fix some bugs of write

ruiqurm committed:
fix bugs of RDMA reading

ruiqurm committed:
fix some bugs and add `MSN`

ruiqurm committed:
fix bugs of MSN

ruiqurm committed:
fix some clippy warnings

ruiqurm committed:
Remove some `unwrap`

ruiqurm committed:
fix some bugs of pervious scheduler

ruiqurm committed:
fix some errors in software

ruiqurm committed:
fix bug of ICRC error and wrong IP

ruiqurm committed:
pass all unit tests

ruiqurm committed:
fix some bugs in software.

ruiqurm committed:
clear all pedantic warnings

ruiqurm committed:
fix some warnings

ruiqurm committed:
remove unreachable_pub

ruiqurm committed:
fix unwrap warnings

ruiqurm committed:
tmp commit

ruiqurm committed:
clean the warning of slicing

ruiqurm committed:
add emulator to software test

ruiqurm committed:
add the full test of emulated to software

ruiqurm committed:
clear arithmetic_side_effects

ruiqurm committed:
clean doc and arithmetic warnings

ruiqurm committed:
handle joining thread

ruiqurm committed:
remove some `FIXME`

ruiqurm committed:
fix bug: a typo about

ruiqurm committed:
add hardware implementation

ruiqurm committed:
add hardware implementation

ruiqurm committed:
fix some bugs of syscall and mmap

ruiqurm committed:
debug physical address resolving

myrfy001 commented on src/lib.rs:
```diff
@@ -1,51 +1,583 @@
+    write_op_ctx_map: Arc<RwLock<HashMap<Msn, WriteOpCtx>>>,
+    ctrl_op_ctx_map: Arc<RwLock<HashMap<u32, CtrlOpCtx>>>,
+    next_ctrl_op_id: AtomicU32,
+    next_msn: AtomicU16,
```
should msn be per device or per QP?

ruiqurm replied:
Yes. We can combine the dqpn with MSN as an u64 , and add a single field in QP context.

ruiqurm replied:
Unfortunately, I found that it seems the MSN field in ETH is not long enough to hold both QP and MSN. We can make some compromises. For example , assign 12 bits as QP number and the rest(4 bits) as MSN per QP. Do we need that?

myrfy001 commented on src/utils.rs:
```diff
@@ -0,0 +1,107 @@
+    pmtu - offset
+}
+
+#[allow(clippy::arithmetic_side_effects)] // total_len must be larger than first_pkt_len
```
the commnt should be greater or equal.

ruiqurm replied:
Yes. I will fix it.

myrfy001 commented on src/utils.rs:
```diff
@@ -0,0 +1,107 @@
+#[inline]
+#[allow(clippy::arithmetic_side_effects)]
+pub(crate) fn get_first_packet_max_length(va: u64, pmtu: u32) -> u32 {
+    // The offset is smaller than pmtu,which is smaller than 4096 currently.
```
add a space after comma

myrfy001 commented on src/device/ringbuf.rs:
```diff
@@ -0,0 +1,421 @@
+        head == tail
+    }
+
+    pub(crate) fn wrapping_add(cur: usize, cnt: usize) -> usize {
```
the algroithm used here is different from what hardware use, so it can't work when the pointer wrap around.
please refer to https://github.com/myrfy001/blue-rdma/blob/3c0fd9f66b04c2333c658b794344c261d004efe3/src/third_party/Ringbuf.bsv#L22 for detail

ruiqurm replied:
OK. I have re-implemented the `ringbuf` using hardware way.

ruiqurm committed:
fix some problems according to pr review

ruiqurm committed:
add `mlock` to lock the huge page

ruiqurm committed:
add `mlock` to lock the huge page

rogercloud commented on src/pd.rs:
```diff
@@ -0,0 +1,96 @@
+    /// # Errors
+    ///
+    /// Will return `Err` if:
+    /// * lock poisoned
```
Do you really care about the lock poison? Does this driver tolerant thread panic?

rogercloud replied:
If not, please consider `parking-lot` instead.

ruiqurm replied:
Sure, I'll take a look at that.

rogercloud commented on src/pkt_checker.rs:
```diff
@@ -0,0 +1,206 @@
+}
+
+struct PacketCheckerContext {
+    send_queue: Sender<RespCommand>,
```
Consier `flume` to replace std channels.

ruiqurm replied:
Sure, I'll take a look at that.

rogercloud commented on src/pkt_checker.rs:
```diff
@@ -0,0 +1,206 @@
+
+struct PacketCheckerContext {
+    send_queue: Sender<RespCommand>,
+    recv_pkt_map: Arc<RwLock<HashMap<Msn, Arc<Mutex<RecvPktMap>>>>>,
```
Types like this is too complecated, please put them in to structure and abstract some repeated operations into functions. For instance, here you code too much to handle lock things.

ruiqurm replied:
Yes. There are amount of patterns like `Arc<RwLock<HashMap<K,V>>>`, I will use a type alias. And for struct like `Arc<Mutex<RecvPktMap>>`, I will create an inner struct and wrap it with Arc and Mutex

rogercloud commented on src/pkt_checker.rs:
```diff
@@ -0,0 +1,206 @@
+        self.stop_flag.store(true, Ordering::Relaxed);
+        if let Some(thread) =  self.thread.take(){
+            if let Err(e) = thread.join(){
+                error!("Failed to join the WorkDescPoller thread: {:?}", e);
```
Thread join returns error when the thread being waited panics, is it enough to logging an error only?

ruiqurm replied:
After reading the docs, I found I may use `.expect("Couldn't join on the associated thread");` because the error is passing to a `panic`.

rogercloud commented on src/pkt_checker.rs:
```diff
@@ -0,0 +1,206 @@
+        };
+        let stop_flag = Arc::new(AtomicBool::new(false));
+        let thread_stop_flag = Arc::clone(&stop_flag);
+        let thread = std::thread::spawn(move || {
```
Why are you creating an thread here? I think a task is enough if you're using an multi-thread async runtime lib.

ruiqurm replied:
I agree with you. To be honest, I am not so familiar with async. So I did not try the async method at the beginning. I will try the async way if there is not too many changes in this pr.

rogercloud commented on src/pkt_checker.rs:
```diff
@@ -0,0 +1,206 @@
+    fn working_thread(ctx: &Self, stop_flag: &AtomicBool) {
+        while !stop_flag.load(Ordering::Relaxed) {
+            if let Err(e) = ctx.check_pkt_map() {
+                error!("PacketChecker stopped: {:?}", e);
```
I prefer "PacketChecker is stopped due to XXXX". And then return directly.

rogercloud commented on src/pkt_checker.rs:
```diff
@@ -0,0 +1,206 @@
+                error!("PacketChecker stopped: {:?}", e);
+                break;
+            }
+        }
```
Here please also log that the worker is stopped by the `stop_flag`.

ruiqurm replied:
I will add that.

rogercloud commented on src/pkt_checker.rs:
```diff
@@ -0,0 +1,206 @@
+                self.send_queue
+                    .send(command)
+                    .map_err(|_| Error::PipeBroken("packet checker send queue"))?;
+                panic!("send nack command")
```
What's the meaning of this panic message? Does it tells the reason of this panic? It just tells where the panic happens.

ruiqurm replied:
In the previous discussion, we agreed that the first version does not need to implement NACK, and we will panic when NACK needs to be sent. I will add a comment here.

rogercloud commented on src/pkt_checker.rs:
```diff
@@ -0,0 +1,206 @@
+    use super::PacketChecker;
+
+    #[test]
+    fn test_packet_checker() {
```
I recommend also check the error case of the functions above.

ruiqurm replied:
I have added one. But more needs to be added for other unit tests. I will keep adding.

rogercloud commented on src/mr.rs:
```diff
@@ -0,0 +1,499 @@
+
+#[derive(Debug)]
+pub(crate) struct MrCtx {
+    #[allow(unused)]
```
Please comment the reason of this unused code, so that we can tell when to remove it later.

ruiqurm replied:
Sure, I will do that.

rogercloud commented on src/mr.rs:
```diff
@@ -0,0 +1,499 @@
+        let mut mr_pgt = self
+            .0
+            .mr_pgt
+            .lock()
```
Again, the lock issue. I rethink about the lock verbose code issue, and find out that you'd better wrap it into a function for a data structure, returning the lock guard instead.

ruiqurm replied:
I am not sure if I understand your meaning. Should I put expressions like `self.0,mr_pgt.lock()` into a method?

ruiqurm committed:
integrate with hardware NIC functions and refactor the chunk buffer in `buf`

ruiqurm committed:
integrate with smoltcp

ruiqurm committed:
use `parking_lot` lock instead

ruiqurm committed:
use `flume::unbounded` to replace std mpsc and crsossbeam mpsc

ruiqurm committed:
remove or add comments for unsed fields

ruiqurm committed:
simplify some structs

ruiqurm committed:
add a simple error unit test. More needs to add.

ruiqurm committed:
ringbuf can select using huge page or aligned memory

ruiqurm committed:
add basic nic utilites

ruiqurm committed:
debug net utilities

ruiqurm committed:
remove some unused imports
